File: 2024/Julia/day02.jl
reports = split.(readlines("R/inst/input02.txt"))

function is_safe(x) 
    v = diff(parse.(Int, x));
    all(Base.between.(v, 1, 3)) | all(Base.between.(v, -3, -1))
end

# Part 1
sum(is_safe.(reports))

# leave one out
function loo(x) 
    vars = [];
    map(i -> push!(vars, deleteat!(deepcopy(x), i)), eachindex(x));
    vars
end

function is_safe_2(x) 
    any(is_safe.(loo(x)))
end

# Part 2
sum(is_safe_2.(reports))

File: 2024/Haskell/day02.hs
import Data.List

slow xs = all ((>= 1) . abs) xs && all ((<= 3). abs) xs

monotonic xs 
    | all (>0) xs = True
    | all (<0) xs = True
    | otherwise = False

isSafe xs = monotonic diffs && slow diffs 
  where diffs = zipWith (-) xs (drop 1 xs)

damped line = zipWith (++) (inits line) (drop 1 $ tails line)

main = do
    p <- readFile "../R/inst/input02.txt"
    let reports = map (map readInt) $ map words $ lines p
    print $ length $ filter isSafe reports -- part 1
    print $ length $ filter ((any isSafe) . damped) reports -- part 2

readInt :: String -> Int
readInt = read 
